docs/user_manual.md

```diff
@@ -349,11 +349,96 @@ this way; it would be an error to apply `symbol(foo)` to another production in
 the module above. Additionally, `symbol(_)` with an argument may not co-occur
 with the `klabel(_)` attribute (see below).
 
+### `overload` attribute
+
+K supports _subsort overloading_[^maude-overload] on symbols, whereby a 
+constructor can have a more specific sort for certain arguments. For example,
+consider the following productions derived from a C-like language semantics:
+```k
+syntax Exp  ::= LVal
+              | Exp  "." Id
+syntax LVal ::= LVal "." Id
+```
+
+Here, it is useful for the result of the dot operator to be an `LVal` if the
+left-hand side is itself an `LVal`. However, there is an issue with the code
+as written: if `L()` is a term of sort `LVal`, then the program `L() . x` has a
+parsing ambiguity between the two productions for the dot operator. To resolve
+this, we can mark the productions as _overloads_:
+```k
+syntax Exp  ::= LVal
+              | Exp  "." Id [overload(_._)]
+syntax LVal ::= LVal "." Id [overload(_._)]
+```
+
+Now, the parser will select the _most specific_ overloaded production when it
+resolves ambiguities in `L() . x` (that is, `L() . x` parses to a term of sort
+`LVal`.
+
+Formally, the compiler organises productions into a partial order that defines
+the overload relation as follows. We say that `P` is a more specific overload
+of `Q` if:
+* `P` and `Q` have the same `overload(_)` attribute. Note that the argument
+  supplied has no semantic meaning other than as a key grouping productions 
+  together.
+* Let `S_P` be the sort of `P`, and `S_p1` etc. be the sorts of its arguments
+  (c.f. for `Q`). The tuple `(S_P, S_p1, ..., S_pN)` must be lexicographically
+  _strictly less than_ `(S_Q, S_q1, ..., S_qN)` according to the definition's
+  subsorting relationship. That is, a term from production `P` must be
+  substitutable where one from `Q` is expected, but cannot be identically
+  sorted.
+
+Care should be taken when the partial order defined by this relationship is not
+also a total order. For example, consider the following
+definition:[^overload-issue]
+```k
+module NUMBERS-SYNTAX
+    imports INT-SYNTAX
+    imports FLOAT-SYNTAX
+
+    syntax IValType ::= "i32" | "i64"
+    syntax FValType ::= "f32" | "f64"
+    syntax AValType ::= IValType | FValType
+    syntax Number ::= Int | Float
+    syntax IVal ::= "<" IValType ">" Int    [overload(<_>_)]
+    syntax FVal ::= "<" FValType ">" Float  [overload(<_>_)]
+    syntax  Val ::= IVal | FVal
+    // the problematic constructor:
+    syntax  Val ::= "<" AValType ">" Number [overload(<_>_)]
```

Comment:
I don't quite follow this example - the `syntax  Val ::= "<" AValType ">" Number` constructor is indeed overly-permissive, but that is always true regardless of any overloading.

Even with a total order, you can still construct similarly problematic examples:
```
module NUMBERS-SYNTAX
    imports INT-SYNTAX
    imports FLOAT-SYNTAX
    syntax SubValType ::= "sub"
    syntax SuperValType ::= "super" | SubValType

    syntax Sub ::= "subVal"
    syntax Super ::= "superVal" | Sub

    syntax SubVal ::= "<" SubValType ">" Sub       [overload(<_>_)]
    syntax SuperVal ::= "<" SuperValType ">" Super [overload(<_>_)] | SubVal
endmodule

module NUMBERS
    imports NUMBERS-SYNTAX
    syntax SuperVal ::= creature32 ( Super ) [function, total]
    rule creature32(S:Sub) => <super> S
    rule creature32(S::Super) => <sub> S
endmodule
```

To me, the over-permissiveness of `syntax  Val ::= "<" AValType ">" Number` is the whole motivation behind overloading in the first place - we want more precise sorting invariants than what K's syntax declaration are able to express.

Still choosing to include the overly-permissive constructor even after using overloading just seems like an odd design decision in the example rather than something fundamental we need to explain to users about overloading. 

Reading through runtimeverification/hs-backend-booster#492, it was mentioned that the problematic constructor was useful because it allows us to write unambiguous rules which handle all overloads like
```
syntax Bar ::= bar(Number) [function, total]
               |  foo(Val) [function, total]
rule foo(< _ > N) => bar(N) // ambiguous without the problematic constructor
```
but I think a more principled and only slightly-inconvenient solution would be to just manually define eliminators
```
syntax AValType ::= valType(Val) [function, total]
rule valType(< V:IValType > _) => V
rule valType(< V:FValType> _ ) => V

syntax Number ::= number(Val) [function, total]
rule number(< _ > N:Int) => N
rule number(< _ > N:Float) => N
```
and write
```
rule foo(V) => bar(number(V))
```

Reply:
Your analysis all makes sense to me; I guess I read Jost's writeup as a cautionary example that needed to be explained to users, and without thinking it through properly made the incorrect inference that because a non-total overload ordering is prone to producing this kind of broken case, total orderings cannot. I guess we can just cut this section from the manual entirely?